Adding an `excludeBeanAttributeNames` configuration

brandonstanleyappfolio opened the pull request:
This change adds a configuration which allows users to specify an MBean attribute exclusion list. This is required when the `blacklistObjectNames` or `whitelistObjectNames` are not sufficient to exclude specific attributes. For our use case, we found it necessary when using the Jmx Exporter for Debezium. There are a couple of metrics which return strings for every table within a connector. Due to the volume of tables, we ran into issues where the JMX exporter would take >10 mins to return each time it was invoked. The metrics that caused this are:

- `CapturedTables` (https://debezium.io/documentation/reference/1.9/connectors/mysql.html#connectors-snaps-metric-capturedtables_mysql)
- `RowsScanned` (https://debezium.io/documentation/reference/1.9/connectors/mysql.html#connectors-snaps-metric-rowsscanned_mysql)

We introduced a new key in the configuration file called `excludeBeanAttributeNames` which is a list of maps where each map contains the bean `ObjectName` and a list of attributes to exclude:

```
excludeBeanAttributeNames:
 - name: <beanObjectName_1>
   values:
    - <METRIC_1_TO_EXCLUDE>
    - <METRIC_2_TO_EXCLUDE>
    - <METRIC_3_TO_EXCLUDE>
    ...
 - name: <beanObjectName_2>
   values:
   - <METRIC_1_TO_EXCLUDE>
   - <METRIC_2_TO_EXCLUDE>
   - <METRIC_3_TO_EXCLUDE>
   ...
```

Using the configuration above, the following steps have been added as part of this change:

1. If there is a key called `excludeBeanAttributeNames` in the yamlConfig, we create a HashMap, where the key is a bean `ObjectName` and the value is a set of attributes to exclude.
2. This HashMap is passed into the constructor for the `JmxScraper`, where we loop over each entry in the HashMap and query the MBeans for each `ObjectName`. We create a new HashMap, where the key is the MBean name, and the value is the set of attributes to exclude.
3. During the loop where we are iterating over each bean to scrape, we check to see if there is an entry in the HashMap for the MBean containing a set of attributes to exclude. If there is, it is passed into the invocation of `scrapeBean`.
4. Within the `scrapeBean` function, while looping over each attribute for a MBean, we skip over any attributes in the set of attributes to exclude.

dhoard commented:
Looks like a useful feature, though I would suggest changing the configuration format. Dynamic keys seem inconsistent with the current configuration format.

```
excludeBeanAttributeNames:
  - name: <beanObjectName_1>
    values:
      - <METRIC_1_TO_EXCLUDE>
      - <METRIC_2_TO_EXCLUDE>
      - <METRIC_2_TO_EXCLUDE>
  - name: <beanObjectName_2>
    values:
      - <METRIC_1_TO_EXCLUDE>
      - <METRIC_2_TO_EXCLUDE>
      - <METRIC_2_TO_EXCLUDE>
```

dhoard commented:
@brandonstanleyappfolio sure.

In your proposal, the keys `<beanObjectName_2>` are dynamic... i.e. bean specific.

I you look at the rules configuration...

```
rules:
  - pattern: 'org.apache.cassandra.metrics<type=(\w+), name=(\w+)><>Value: (\d+)'
    name: cassandra_$1_$2
    value: $3
```

The rule `name` contains the name of the rule, and `value` contains the value of the rule.

Essentially, make the key and value constants. (`excludeBeanAttributeNames` is map of maps, where the inner map has 2 keys `name` = String, and `value`= array of String.

I seem to find users misconfiguring systems with the approach you suggest...just my opinion.

brandonstanleyappfolio commented:
@dhoard I've added the change for the new configuration format and updated the `README`, commit message, and PR description. Can you take a look when you have a chance?

brandonstanleyappfolio commented:
Hey @dhoard! Do you know if there is a process that I can follow to get this change reviewed?

dhoard commented:
@brandonstanleyappfolio there are a few unresolved comments.

My current priority is to get the integration test suite integrated so that we can do comprehensive testing.

dhoard commented:
> > @brandonstanleyappfolio there are a few unresolved comments.
> > My current priority is to get the integration test suite integrated so that we can do comprehensive testing.
> 
> Can you clarify which comments are unresolved? If you're referring to [this](https://github.com/prometheus/jmx_exporter/pull/781#issuecomment-1474148072) comment,I've incorporated these changes, as per my comment [here](https://github.com/prometheus/jmx_exporter/pull/781#issuecomment-1474509579).

Apologies... the review changes were pending.

brandonstanleyappfolio commented on collector/src/main/java/io/prometheus/jmx/JmxScraper.java:
```diff
@@ -145,6 +159,11 @@ private void scrapeBean(MBeanServerConnection beanConn, ObjectName mbeanName) {
                 logScrape(mbeanName, attr, "not readable");
                 continue;
             }
+
```
I believe this would not work because we need to know which mBeans contain the excluded attributes. The `name` field within each `excludeBeanAttributeNames` record is "generic" since it can contain wildcards and other regex patterns. 

For example, suppose we have the following specified for `excludeBeanAttributeNames`:

```
excludeBeanAttributeNames:
 - name: "debezium.mysql:type=*,context=*,server=*"
   values:
    - TotalNumberOfEventsSeen
```
The `if (excludeBeanAttributeNames.contains(mbeanName) {` check would not work because `mbeanName` would contain the substituted regex patterns with actual values, like: `debezium.mysql:type=connector-metrics,context=streaming,server=dbserver1`. Please let me know if my understanding is incorrect.

brandonstanleyappfolio replied:
@dhoard

dhoard replied:
@brandonstanleyappfolio

Given MBeans...

```
debezium.mysql:type=connector-metrics,context=streaming,server=dbserver1
debezium.mysql:type=connector-metrics,context=streaming,server=dbserver2
debezium.mysql:type=connector-metrics,context=streaming,server=dbserver3
```

And an `excludeBeanAttributeNames` pattern...

```
excludeBeanAttributeNames:
 - name: "debezium.mysql:type=*,context=*,server=server1"
   values:
    - TotalNumberOfEventsSeen
```

In the scenario above, we want to only exclude `TotalNumberOfEventsSeen` for the specific Object name that matches `debezium.mysql:type=*,context=*,server=server1`

While this is odd, it could happen for various reasons.

Can you add a test case with multiple MBeans and the above `excludeBeanAttributeNames` scenario?

brandonstanleyappfolio replied:
@dhoard Yep - added! Here is unit test: https://github.com/prometheus/jmx_exporter/blob/20977c1d11d8bb42e1c417aab4e57304d543b970/collector/src/test/java/io/prometheus/jmx/JmxCollectorTest.java#L373

dhoard commented on collector/src/main/java/io/prometheus/jmx/JmxScraper.java:
```diff
@@ -110,12 +113,23 @@ public void doScrape() throws Exception {
                 }
             }
 
+            Map<ObjectName, HashSet<String>> attributesToExcludeFromBeans = new HashMap<ObjectName, HashSet<String>>();
+            for (Map.Entry<ObjectName, HashSet<String>> entry : excludeBeanAttributeNames.entrySet()) {
+                for (ObjectInstance instance: beanConn.queryMBeans(entry.getKey(), null)) {
+                    attributesToExcludeFromBeans.put(instance.getObjectName(), entry.getValue());
+                }
+            }
+
```
This could cause a performance issue when using the exporter in standalone mode (and possibly agent mode.)

Add better approach would be to query the beans, then when processing through the attributes, filter out any attributes on the exclude list.

brandonstanleyappfolio commented:
Regarding [this](https://github.com/prometheus/jmx_exporter/pull/781#discussion_r1174177874), do you know if there a method I can use to look up an `MBean` name against the regex list provided in the YAML configuration file?

dhoard commented:
Not that I'm aware of. 

My understanding is the feature is to excluded specific MBean attributes from collection.

You should be able to filter the attributes during processing of the MBeans.

i.e.

If the MBean ObjectName is in the exclude Map, get and check exclude attribute set.

If the attribute is in the exclude attribute set, skip it, otherwise process it.

Am I missing something?

brandonstanleyappfolio commented:
Given the following `MBeans`:

```
debezium.mysql:type=connector-metrics,context=streaming,server=dbserver1
debezium.mysql:type=connector-metrics,context=streaming,server=dbserver2
debezium.mysql:type=connector-metrics,context=streaming,server=dbserver3
```

and the following `excludeBeanAttributeNames` configuration:

```
excludeBeanAttributeNames:
 - name: "debezium.mysql:type=*,context=*,server=*"
   values:
    - TotalNumberOfEventsSeen
```

How would we check if the `MBeans` match an entry from the `excludeBeanAttributeNames` map? 

The MBeans are specific (**Example**: `type=connector-metrics`, `context=streaming`, and `server=dbserver1`), but the `excludeBeanAttributeNames` are not and instead contain regex patterns (**Example**: `type=*`, `context=*`, and `server=*`). It could be possible to specify every `MBean` in the `excludeBeanAttributeNames` map, but this approach cannot be used unless you know all possible values beforehand (**Example**: Our use case has over 1000 values for the `server` component of the `MBean`, which are dynamic and are unknown in advance).

dhoard commented:
The exclude definition would be a regular expression to match an MBean (pattern), then a attribute name (or possibly another regular expression to match the attribute name.)

dhoard commented:
I would suggest a regular expression (pattern) for the attribute names as it's more flexible.

dhoard commented:
While I make it sound simple... I know it's a lot of work. The cod to filter would probably have to be in https://github.com/prometheus/jmx_exporter/blob/6f95075b4ea7965a9fc446b9cdc951cdf43e2646/collector/src/main/java/io/prometheus/jmx/JmxScraper.java#L210 along with passing the MBean `ObjectName`

brandonstanleyappfolio commented:
Yes, I see why this could be more efficient, though. With the approach I've built, we list all `MBeans` again to determine which ones to exclude the attributes from. So it makes sense to filter out the attributes in the `processBeanValue` function. I will look into updating the method for this feature.

dhoard commented:
Closing. Functionality merged for the 0.20.0 release.
